### PR TITLE
Add a docstring & tests for _flatten_dict.

### PR DESCRIPTION
**Pull request by clokep**
I pulled this out of #14964 since I want them for another PR (and to make that PR slightly smaller).

### CODE REVIEW

File: synapse/push/bulk_push_rule_evaluator.py [hunk 1]
```diff
@@ -466,6 +466,27 @@ def _flatten_dict(
...
+    Given a JSON dictionary (or event) which might contain sub dictionaries,
+    flatten it into a single layer dictionary by combining the keys & sub-keys.
+
+    Transforms:
+
+        {"foo": {"bar": "test"}}
+
+    To:
+
+        {"foo.bar": "test"}
```
Comment (DMRobertson) [inline synapse/push/bulk_push_rule_evaluator.py:479]: How would (should) the dict 

```python
{"foo.bar": 1, "foo": {"bar": 2}}
```

be transformed? I imagine that `2` will win... but maybe we should just say something like "the result is undefined if there is any kind of conflict between keys"?

File: tests/push/test_push_rule_evaluator.py [hunk 2]
```diff
@@ -37,6 +37,30 @@
 from tests.test_utils.event_injection import create_event, inject_member_event
 
 
+class FlattenDictTestCase(unittest.TestCase):
```
Comment (DMRobertson) [inline tests/push/test_push_rule_evaluator.py:40]: Might be worth a test case where this function is given an `EventBase`---it seems to have bespoke logic for events.

File: synapse/push/bulk_push_rule_evaluator.py [hunk 3]
```diff
@@ -466,6 +466,27 @@ def _flatten_dict(
     prefix: Optional[List[str]] = None,
     result: Optional[Dict[str, str]] = None,
 ) -> Dict[str, str]:
+    """
+    Given a JSON dictionary (or event) which might contain sub dictionaries,
+    flatten it into a single layer dictionary by combining the keys & sub-keys.
+
```
Comment (reivilibre) [inline synapse/push/bulk_push_rule_evaluator.py:472]: Should note that non-string items are dropped; doesn't sound obvious from the name

**Review by reivilibre — APPROVED** [review]

File: synapse/push/bulk_push_rule_evaluator.py [hunk 4]
```diff
@@ -466,6 +466,27 @@ def _flatten_dict(
     prefix: Optional[List[str]] = None,
     result: Optional[Dict[str, str]] = None,
 ) -> Dict[str, str]:
+    """
+    Given a JSON dictionary (or event) which might contain sub dictionaries,
+    flatten it into a single layer dictionary by combining the keys & sub-keys.
```
Comment (DMRobertson) [inline synapse/push/bulk_push_rule_evaluator.py:471]: Can we mention that non-string (leaf) values are dropped?

**Review by DMRobertson — COMMENTED** [review]
Many thanks, with apologies for pedantry

File: synapse/push/bulk_push_rule_evaluator.py (hunk 1 above)
Comment (clokep) [reply synapse/push/bulk_push_rule_evaluator.py:479]: This is undefined behavior, one of which the pull requests that I'll be implementing will be fixing.

File: tests/push/test_push_rule_evaluator.py (hunk 2 above)
Comment (clokep) [reply tests/push/test_push_rule_evaluator.py:40]: Maybe -- I'm not changing that behavior so I'll try to do that as a follow-up so I can at least avoid the self-conflicts.

**Review by DMRobertson — APPROVED** [review]

File: synapse/push/bulk_push_rule_evaluator.py (hunk 1 above)
Comment (clokep) [reply synapse/push/bulk_push_rule_evaluator.py:481]: MSC3873 for the record.

**Pull request merged**